### PR TITLE
Linter refactor

### PR DESCRIPTION
**Pull request by shellscape**
All commits from the last 20 days have been reintegrated into the refactor branch.

We still need to squash merge this, but for the purposes of review, I'm leaving all of the reintegration commits in the history for now. We'll squash merge before merging the PR. We'll also likely have to force the PR onto master.

### CODE REVIEW

**Comment by shellscape** [comment]
merged to master and resolved the conflicts.

`npm test` passes. this one should be ready for manual merging tomorrow.

File: .jscsrc [hunk 1]
```diff
@@ -4,8 +4,9 @@
     "maximumLineLength": null,
 
     "disallowMultipleVarDecl": "strict",
-    "disallowSpacesInAnonymousFunctionExpression": null,
-    "disallowSpacesInFunctionExpression": null,
+    "disallowSpacesInAnonymousFunctionExpression": false,
```
Comment (jwilsson) [inline .jscsrc]: I think all of these options needs to be `null`, the Appveyor build is failing on this (it passes locally and on Travis though, which is super weird).

File: lib/linter.js [hunk 2]
```diff
@@ -75,6 +75,7 @@ exports.resultSeverity = {
 exports.lint = function (input, fullPath, config) {
     var results = [];
     var ast;
+    var extend = require('lodash.assign');
```
Comment (jwilsson) [inline lib/linter.js]: I really think we should just use `lodash.merge()` instead. It seems really unnecessary to include a second (almost identical) dependency. Could we also keep the `require()` calls grouped at the top?

File: lib/linter.js [hunk 3]
```diff
@@ -83,14 +84,45 @@ exports.lint = function (input, fullPath, config) {
...
+            }
+
+            if (linter.nodeTypes && Array.isArray(linter.nodeTypes) && linter.nodeTypes.length) {
+                // Not applicable, bail
+                if (linter.nodeTypes.indexOf(node.type) < 0) {
+                    return;
+                }
+            }
+
+            lint = linter.lint.call(linter, config, node);
```
Comment (jwilsson) [inline lib/linter.js]: Is the `.call()` really needed?

File: lib/linters/attribute_quotes.js [hunk 4]
```diff
@@ -1,46 +1,35 @@
...
-    }
+    name: 'attributeQuotes',
+    nodeTypes: ['simpleSelector'],
+    message: 'Attribute selectors should use quotes.',
 
-    selector = node.first('attribute');
+    lint: function attributeQuotesLinter (config, node) {
+        var selector;
+        var value;
+        var sprintf = require('sprintf-js').sprintf;
```
Comment (jwilsson) [inline lib/linters/attribute_quotes.js]: Couldn't we use Node core `util.format()` instead?

File: test/lib/spec_linter.js [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+
+        var results = [];
+        var extend = require('lodash.assign');
+
+        // Bail if the linter isn't wanted
+        if (!config[linter.name] || (config[linter.name] && !config[linter.name].enabled)) {
+            return;
+        }
+
+        if (linter.nodeTypes && Array.isArray(linter.nodeTypes) && linter.nodeTypes.length) {
```
Comment (jwilsson) [inline test/lib/spec_linter.js]: I think we could remove the `length` check, `indexOf` will return `-1` on an empty array anyway,

File: test/lib/spec_linter.js [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+            if (linter.nodeTypes.indexOf(node.type) < 0) {
+                return;
+            }
+        }
+
+        lint = linter.lint.call(linter, config, node);
+
+        if (lint) {
+            // transform the linter result into an array
+            // some linters return an array, some don't. and that's OK.
```
Comment (jwilsson) [inline test/lib/spec_linter.js]: Doesn't all linters return an array now? So we can remove this comment?

File: lib/linters/duplicate_property.js [hunk 7]
```diff
@@ -1,54 +1,49 @@
...
-                message: result.message
-            };
+        node.forEach('declaration', function (declaration) {
+            var property = declaration.first('property');
+
+            property = helpers.ensureObject(property.first('ident'));
+
+            if (properties.indexOf(property.content) !== -1) {
+                results.push({
+                    // `this.message` works since we're allowed access to ourself as a linter object
```
Comment (jwilsson) [inline lib/linters/duplicate_property.js]: Unnecessary comment.

File: lib/linters/duplicate_property.js [hunk 8]
```diff
@@ -1,54 +1,49 @@
...
+
+            if (excludes.indexOf(property.content) === -1) {
+                properties.push(property.content);
+            }
         });
-    }
 
-    return null;
+        if (results.length) {
+            // we moved the 'linter' prop to linter.js,
```
Comment (jwilsson) [inline lib/linters/duplicate_property.js]: Unnecessary comment.

File: lib/linters/duplicate_property.js [hunk 9]
```diff
@@ -1,54 +1,49 @@
...
-        var property = declaration.first('property');
+    lint: function duplicatePropertyLinter (config, node) {
+        var helpers = require('../helpers');
+        var properties = [];
+        var excludes = [];
+        var results = [];
+        var self = this;
 
-        property = helpers.ensureObject(property.first('ident'));
+        // if we see this being used frequently, we can introduce a `util` param to the lint function
```
Comment (jwilsson) [inline lib/linters/duplicate_property.js]: Unnecessary comment, I think we can leave this as-is for now.

File: lib/linters/space_after_property_colon.js [hunk 10]
```diff
@@ -1,57 +1,55 @@
...
+    lint: function spaceAfterPropertyColonLinter (config, node) {
+
+        var checkIndex;
+        var maybeSpace;
+        var findIndex = require('lodash.findindex');
+        var sprintf = require('sprintf-js').sprintf;
+        var style = config.spaceAfterPropertyColon.style;
+        var valid = true;
+
+        // Find the colon (south of the spleen)
```
Comment (jwilsson) [inline lib/linters/space_after_property_colon.js]: Clever ;)

File: lib/linters/space_after_property_name.js [hunk 11]
```diff
@@ -1,57 +1,54 @@
...
+    lint: function spaceAfterPropertyNameLinter (config, node) {
+
+        var checkIndex;
+        var maybeSpace;
+        var findIndex = require('lodash.findindex');
+        var sprintf = require('sprintf-js').sprintf;
+        var style = config.spaceAfterPropertyName.style;
+        var valid = true;
+
+        // Find the colon (it's next to the prostate)
```
Comment (jwilsson) [inline lib/linters/space_after_property_name.js]: Clever ;)

File: test/specs/linters/attribute_quotes.js [hunk 12]
```diff
@@ -61,34 +57,29 @@ describe('lesshint', function () {
...
-                config: options,
-                node: ast,
-                path: 'test.less'
-            });
+            actual = lint(options, ast);
 
             assert.deepEqual(actual, expected);
         });
 
         it('should return null for value-less selectors', function () {
```
Comment (jwilsson) [inline test/specs/linters/attribute_quotes.js]: Change all instances of `should return null [...]` to `should return undefined [...]` (or something like that).

File: lib/linters/decimal_zero.js [hunk 13]
```diff
@@ -1,89 +1,89 @@
...
-    var regexp = /^-?(0|\d+)\.(\d*0)$/;
-    return regexp.test(number);
-};
+    lint: function decimalZeroLinter (config, node) {
 
-var isValidNone = function (number) {
-    var regexp = /^-?([^0]\d*)?\.(\d*[^0])$/;
-    return regexp.test(number);
-};
+        function isValidTrailing (number) {
```
Comment (jwilsson) [inline lib/linters/decimal_zero.js]: I think we should move all of these function definitions outside, bit cleaner IMO.

File: lib/linters/single_line_per_property.js [hunk 14]
```diff
@@ -1,99 +1,86 @@
...
+        if (results.length) {
+            return results;
+        }
+    },
 
-module.exports = function (options) {
-    var config = options.config;
-    var node = options.node;
-    var results = [];
+    isPreviousValid: function (node, element, index) {
```
Comment (jwilsson) [inline lib/linters/single_line_per_property.js]: I don't think we need to export these functions (`isPreviousValid()` and `isNextValid()`), lets move them outside.

File: lib/linters/attribute_quotes.js [hunk 15]
```diff
@@ -1,46 +1,35 @@
...
-    var config = options.config;
-    var node = options.node;
-    var selector;
-    var value;
-
-    // Bail if the linter isn't wanted
-    if (!config.attributeQuotes || (config.attributeQuotes && !config.attributeQuotes.enabled)) {
-        return null;
-    }
+module.exports = {
```
Comment (jwilsson) [inline lib/linters/attribute_quotes.js]: I hate to be this nitpicky but the empty lines after each of the `module.exports` and function definitions shouldn't be there, for consistency with other parts of the code. `"disallowPaddingNewlinesInBlocks": true` in `.jshintrc` can help catch some of them.

**Comment by jwilsson** [comment]
First of all, awesome job! A huge thanks for tackling this and sorry about the merge mess (I'll take some time squashing everything, fixing conflicts etc.).

Some nitpicks, mostly stylistic issues. Let me know if you want me to handle some of it.

Again, a huge thank you!

File: lib/linter.js [hunk 16]
```diff
@@ -83,14 +84,45 @@ exports.lint = function (input, fullPath, config) {
...
-        }
+        linters.forEach(function (linter) {
+            var lint;
+
+            // Bail if the linter isn't wanted
+            if (!config[linter.name] || (config[linter.name] && !config[linter.name].enabled)) {
+                return;
+            }
+
+            if (linter.nodeTypes && Array.isArray(linter.nodeTypes) && linter.nodeTypes.length) {
```
Comment (jwilsson) [inline lib/linter.js]: I think we could remove the `length` check, `indexOf` will return `-1` on an empty array anyway,

File: lib/linter.js [hunk 17]
```diff
@@ -83,14 +84,45 @@ exports.lint = function (input, fullPath, config) {
...
+                if (linter.nodeTypes.indexOf(node.type) < 0) {
+                    return;
+                }
+            }
+
+            lint = linter.lint.call(linter, config, node);
+
+            if (lint) {
+                // transform the linter result into an array
+                // some linters return an array, some don't. and that's OK.
```
Comment (jwilsson) [inline lib/linter.js]: Doesn't all linters return an array now? So we can remove this comment?

File: lib/linters/decimal_zero.js (hunk 13 above)
Comment (shellscape) [inline lib/linters/decimal_zero.js]: I moved them bc I was trying to keep the header of the file uniform. Working on all 20-something linters at once, it was easier to see a uniform start/header to every file.

**Comment by shellscape** [comment]
@jwilsson if those are blocking on this PR, please move forward with what you'd like to see changed. I spent just about all the time I had for last week (for lesshint) and this week into the refactor, and I probably wont be able to get to any of those things you'd like to see different for a good week or so. Since this PR is from a branch on lesshint (and not the gilt) fork, please proceed with any of the code style changes you'd like to see in order to get this PR merged.

File: lib/linter.js (hunk 17 above)
Comment (shellscape) [inline lib/linter.js]: yes, indeed

**Comment by jwilsson** [comment]
@shellscape I totally understand, I'm just picky ;)

I'll try to find some time tomorrow, and at least fix the Appveyor build and try to merge it as nice as possible. The other things aren't that important and can be adjusted later.

Again, thanks for tackling this. I really appreciate it.

**Comment by jwilsson** [comment]
@shellscape This is merged in 92bb70a7c53310cfdf00b4ec771d38a1ebee66ff. 

I removed the comments and made sure each linter just get passed the relevant options.

**Comment by shellscape** [comment]
:+1:

**Pull request closed without merging**